File: solutions/day09/day09.hs
import Text.Printf (printf)

differences :: [Int] -> [Int]
differences hist = zipWith subtract hist (tail hist)

predict :: Bool -> [Int] -> Int
predict past hist
  | all (== 0) hist = 0
  | past = head hist - predict past (differences hist)
  | otherwise = last hist + predict past (differences hist)

interpolate :: Bool -> [[Int]] -> Int
interpolate past = sum . map (predict past)

parseInput :: String -> [Int]
parseInput = map read . words

main :: IO ()
main = do
  input <- getContents
  let inp = map parseInput . lines $ input
  printf "%d %d\n" (interpolate False inp) (interpolate True inp)

File: solutions/day09/day09.py
from itertools import pairwise
from functools import partial

inp = [list(map(int, line.split())) for line in open(0).readlines()]


def predict(hist, past=False):
    if all(x == 0 for x in hist):
        return 0

    return hist[0 if past else -1] + (-1 if past else 1) * predict(
        [b - a for a, b in pairwise(hist)], past
    )


def part1():
    return sum(map(predict, inp))


def part2():
    return sum(map(partial(predict, past=True), inp))


print(part1(), part2())
